myia/grad.py

```diff
@@ -0,0 +1,657 @@
+
+        * Sensitivities for all free variables are packed in an
+          EnvInstance using env_setitem.
+        * We return a tuple with fv sensitivities first, and then
+          all parameter sensitivities.
+        """
+        fv_sens = Constant(newenv)
+        for fv in g.free_variables_total:
+            sens = self.get(g, fv)
+            # NOTE: If sens is an application of zeros_like, it would be
+            # possible to skip adding it to the env.
+            fv_sens = ng.apply(
+                primops.env_setitem,
+                fv_sens,
+                ng.apply(primops.embed, self.get_jinv(fv)),
+                sens
+            )
+        in_sens = [self.get(g, p) for p in g.parameters]
+        ng.output = ng.apply(primops.make_tuple,
+                             fv_sens,
+                             *in_sens)
+        if len(ng.parameters) == 0:
+            # This can happen if the output is a constant. In that case we just
+            # add a dummy parameter to satisfy the backpropagator protocol.
+            with About(g.output.debug, 'grad_sens'):
+                ng.add_parameter()
+
+
+class RemapperSet:
+    """Set of remappers working together to generate one or more graphs."""
+
+    def __init__(self, graphs, **remappers):
+        """Initialize a RemapperSet."""
+        self.remappers = {
+            name: remapper(relation=name, graphs=graphs, remappers=self)
+            for name, remapper in remappers.items()
+        }
+        self.graphs = graphs
+
+    def run(self):
+        """Run all remappers.
+
+        All remappers are populated first, then linked, then finalized.
+        """
+        for remapper in self.remappers.values():
+            remapper.populate()
+        for remapper in self.remappers.values():
+            remapper.link()
+        for remapper in self.remappers.values():
+            remapper.finalize()
+
+    def __getitem__(self, item):
+        return self.remappers[item]
+
+
+def _grad(mng, root):
+    graphs = root.scope
+
+    remappers = RemapperSet(
+        graphs,
+        grad_fprop=FPropRemapper.partial(),
+        grad_fprop_app=FPropAppRemapper.partial(
+            master='grad_fprop'
+        ),
+        grad_bprop=BPropRemapper.partial(
+            master='grad_fprop'
+        ),
+        grad_sens=SensRemapper.partial(
+            graph_relation='grad_bprop'
+        ),
+        grad_bprop_app=BPropAppRemapper.partial(
+            master='grad_sens'
+        ),
+    )
+    remappers.run()
+    return remappers['grad_fprop'].get_graph(root)
+
+
+@overload
+def J(prim: Primitive, resources):
+    """Implement J on a Primitive."""
+    g = augmented_graphs[prim]
+    if isinstance(g, Graph):
+        return clone(g)
+    else:
+        return g
+
+
+@overload  # noqa: F811
+def J(graph: Graph, manager):
+    """Implement J on a Graph."""
+    if graph.transforms.get('grad', None):
+        return graph.transforms['grad']
```

Comment:
Would that risk to cause perturbation confusion if `J` is called more than once?

Reply:
No, this is just to ensure a graph is only transformed once. Different calls of the result of `J(f)` will compute their perturbations separately.

Reply:
OK.